src/main/java/org/testng/internal/TestResult.java

```diff
@@ -320,4 +336,100 @@ public boolean wasRetried() {
   public void setWasRetried(boolean wasRetried) {
     this.m_wasRetried = wasRetried;
   }
+
+  public List<ITestNGMethod> getSkipCausedBy() {
+    if (this.m_status != SKIP) {
+      return Collections.unmodifiableList(skippedDueTo);
+    }
+    if (skipAnalysed) {
+      return Collections.unmodifiableList(skippedDueTo);
+    }
+    skipAnalysed = true;
+    //check if there were any config failures
+    Set<ITestResult> skippedConfigs = m_context.getFailedConfigurations().getAllResults();
+    for (ITestResult skippedConfig : skippedConfigs) {
+      if (isGlobalFailure(skippedConfig) || isRelated(skippedConfig) ) {
+        //If there's a failure in @BeforeTest/@BeforeSuite/@BeforeClass
+        //then the reason is most often just one method.
+        skippedDueTo.add(skippedConfig.getMethod());
+        return Collections.unmodifiableList(skippedDueTo);
+      }
+      if (belongToSameGroup(skippedConfig)) {
+        //If its @BeforeGroups then there's a chance that there could be more than one
+        //method. So lets add everything.
+        skippedDueTo.add(skippedConfig.getMethod());
+      }
+    }
+    if (!skippedDueTo.isEmpty()) {
+      //If we found atleast one skipped due to reason, then its time to return back.
+      return Collections.unmodifiableList(skippedDueTo);
+    }
+    //Looks like we didnt have any configuration failures. So some upstream method perhaps failed.
+    List<String> upstreamMethods = Arrays.asList(m_method.getMethodsDependedUpon());
+    if (upstreamMethods.isEmpty()) {
+      //Maybe group dependencies exist ?
+      List<String> upstreamGroups = Arrays.asList(m_method.getGroupsDependedUpon());
+      if (upstreamGroups.isEmpty()) {
+        return Collections.unmodifiableList(skippedDueTo);
+      }
+      List<ITestResult> allFailures = Lists.newArrayList(m_context.getFailedTests().getAllResults());
+      allFailures.addAll(m_context.getFailedButWithinSuccessPercentageTests().getAllResults());
+      skippedDueTo = allFailures.stream()
+          .map(ITestResult::getMethod)
+          .filter(each -> {
+            List<String> currentMethodGroups = Arrays.asList(each.getGroups());

```

Comment:
you should be able to add 1 or 2 other map usages here:
```java
skippedDueTo = allFailures.stream()
          .map(ITestResult::getMethod)
          .map(method -> Arrays.asList(method.getGroups())
          .map(groups -> Lists.intersection(upstreamGroups, currentMethodGroups))
          .filter(List::isEmpty)
```

Reply:
No this will not work. The `map()` causes a transformation.

So when you do `.map(method -> Arrays.asList(method.getGroups())`, you have caused the stream to transform from a `Stream<ITestNGMethod>` to a `Stream<List<String>>` which is not useful for me. I would like to filter out methods that satisfy my filtering criteria.

Here's my understanding of these two:
`.filter` --> lets me apply selection criteria to filter out the output from the stream
`.map` --> causes me to apply some conversion logic and transform a stream from one form to the other.

Reply:
Ok, right!